cpp/src/arrow/python/numpy_to_arrow.cc

```diff
@@ -670,9 +671,11 @@ Status NumPyConverter::Visit(const StringType& type) {
         return Status::Invalid("Encountered non-UTF8 binary value: ",
                                HexEncode(data, itemsize_));
       }
-    } else {
+    } else if (is_unicode_type) {
       return AppendUTF32(reinterpret_cast<const char*>(data), itemsize_, byteorder,
                          &builder);
+    } else {
+      return Status::Invalid("Expected a string or bytes dtype, got ", dtype_->kind);
```

Comment:
Is there a way to get a nicer name for the numpy dtype except for its "kind" (from the C API)?

Reply:
You can probably use `PyObject_Str`.

Reply:
You can also call directly one of our helpers, `PyObject_StdStringStr` or `PyObject_StdStringRepr`.

Reply:
Thanks, updated with `PyObject_StdStringStr`